File: javascript/calculadora.js
/**
 * Programa: calculadora.js
 * Descrição: Interface de console interativa para a Calculadora de Matrizes em Node.js.
 * Disciplina: Álgebra Linear — Fatec Osasco
 * Curso: Desenvolvimento de Software Multiplataforma
 */

const readline = require("readline");
const {
  obterDimensoes,
  criarMatrizVazia,
  gerarMatrizAleatoria,
  formatarMatriz,
  somarMatrizes,
  subtrairMatrizes,
  multiplicarEscalar,
  transporMatriz,
  multiplicarMatrizes,
  verificarQuadrada,
  verificarSimetrica,
  verificarIdentidade,
  calcularTraco,
  calcularDeterminante
} = require("./core_matrizes");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

/**
 * Função utilitária assíncrona para solicitar entrada ao usuário.
 * @param {string} query 
 * @returns {Promise<string>}
 */
function ask(query) {
  return new Promise((resolve) => rl.question(query, resolve));
}

/**
 * Lê um número inteiro estritamente positivo com validação.
 * @param {string} mensagem 
 * @returns {Promise<number>}
 */
async function lerInteiroPositivo(mensagem) {
  while (true) {
    const input = (await ask(mensagem)).trim();
    const val = parseInt(input, 10);
    if (!isNaN(val) && val > 0 && String(val) === input) {
      return val;
    }
    console.log(">> Erro: O valor deve ser um número inteiro maior que zero.");
  }
}

/**
 * Lê um número real (float) com validação de entrada.
 * @param {string} mensagem 
 * @returns {Promise<number>}
 */
async function lerFloat(mensagem) {
  while (true) {
    const input = (await ask(mensagem)).trim().replace(",", ".");
    const val = parseFloat(input);
    if (!isNaN(val)) {
      return val;
    }
    console.log(">> Erro: Entrada inválida. Por favor, digite um número real válido.");
  }
}

/**
 * Fluxo guiado para cadastro de matriz (Manual ou Aleatória).
 * @param {string} nome 
 * @returns {Promise<number[][]>}
 */
async function cadastrarMatrizFluxo(nome) {
  console.log(`\n--- Cadastro da Matriz ${nome} ---`);
  const linhas = await lerInteiroPositivo(`Informe o número de linhas de ${nome}: `);
  const colunas = await lerInteiroPositivo(`Informe o número de colunas de ${nome}: `);

  console.log("\nEscolha o método de preenchimento:");
  console.log("  1 - Manual (digitar elemento por elemento)");
  console.log("  2 - Automático (valores inteiros aleatórios)");

  while (true) {
    const opcao = (await ask("Opção desejada [1 ou 2]: ")).trim();

    if (opcao === "1") {
      const matriz = criarMatrizVazia(linhas, colunas);
      console.log(`\nDigite os elementos da matriz ${nome} (${linhas}x${colunas}):`);
      for (let i = 0; i < linhas; i++) {
        for (let j = 0; j < colunas; j++) {
          matriz[i][j] = await lerFloat(`  Elemento [${i + 1}, ${j + 1}]: `);
        }
      }
      console.log(`\nMatriz ${nome} cadastrada com sucesso!`);
      console.log(formatarMatriz(matriz, nome));
      return matriz;
    } else if (opcao === "2") {
      const minInput = (await ask("Valor mínimo para os números aleatórios [padrão -10]: ")).trim();
      const minVal = minInput === "" ? -10 : parseInt(minInput, 10);

      const maxInput = (await ask("Valor máximo para os números aleatórios [padrão 10]: ")).trim();
      const maxVal = maxInput === "" ? 10 : parseInt(maxInput, 10);

      const matriz = gerarMatrizAleatoria(linhas, colunas, minVal, maxVal);
      console.log(`\nMatriz ${nome} gerada aleatoriamente com sucesso!`);
      console.log(formatarMatriz(matriz, nome));
      return matriz;
    } else {
      console.log(">> Opção inválida! Escolha 1 para Manual ou 2 para Automático.");
    }
  }
}

/**
 * Seleciona Matriz A ou B para operações unárias.
 * @param {number[][]|null} matrizA 
 * @param {number[][]|null} matrizB 
 * @returns {Promise<{ matriz: number[][], nome: string } | null>}
 */
async function selecionarMatriz(matrizA, matrizB) {
  console.log("\nQual matriz deseja utilizar?");
  console.log("  A - Matriz A");
  console.log("  B - Matriz B");
  const escolha = (await ask("Escolha [A/B]: ")).trim().toUpperCase();

  if (escolha === "A") {
    if (!matrizA) {
      console.log(">> Atenção: A Matriz A ainda não foi cadastrada.");
      return null;
    }
    return { matriz: matrizA, nome: "A" };
  } else if (escolha === "B") {
    if (!matrizB) {
      console.log(">> Atenção: A Matriz B ainda não foi cadastrada.");
      return null;
    }
    return { matriz: matrizB, nome: "B" };
  } else {
    console.log(">> Opção inválida! Escolha 'A' ou 'B'.");
    return null;
  }
}

/**
 * Loop principal do menu interativo em console.
 */
async function main() {
  let matrizA = null;
  let matrizB = null;

  while (true) {
    console.log("\n" + "=".repeat(45));
    console.log("         CALCULADORA DE MATRIZES");
    console.log("=".repeat(45));
    console.log(" 1  - Cadastrar Matriz A");
    console.log(" 2  - Cadastrar Matriz B");
    console.log(" 3  - Exibir Matriz A");
    console.log(" 4  - Exibir Matriz B");
    console.log(" 5  - Somar (A + B)");
    console.log(" 6  - Subtrair (A - B)");
    console.log(" 7  - Multiplicar (A x B)");
    console.log(" 8  - Multiplicar Matriz por Escalar");
    console.log(" 9  - Transpor Matriz");
    console.log("--- [Opcionais / Bônus] ---");
    console.log(" 10 - Verificar Propriedades (Quadrada, Simétrica, Identidade)");
    console.log(" 11 - Calcular Traço da Matriz");
    console.log(" 12 - Calcular Determinante");
    console.log(" 0  - Sair");
    console.log("=".repeat(45));

    const opcao = (await ask("Digite a opção desejada: ")).trim();

    if (opcao === "1") {
      matrizA = await cadastrarMatrizFluxo("A");

    } else if (opcao === "2") {
      matrizB = await cadastrarMatrizFluxo("B");

    } else if (opcao === "3") {
      if (!matrizA) {
        console.log(">> Matriz A não cadastrada. Utilize a opção 1 primeiro.");
      } else {
        console.log("\n" + formatarMatriz(matrizA, "A"));
      }

    } else if (opcao === "4") {
      if (!matrizB) {
        console.log(">> Matriz B não cadastrada. Utilize a opção 2 primeiro.");
      } else {
        console.log("\n" + formatarMatriz(matrizB, "B"));
      }

    } else if (opcao === "5") {
      if (!matrizA || !matrizB) {
        console.log(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a soma.");
      } else {
        try {
          const resultado = somarMatrizes(matrizA, matrizB);
          console.log("\n--- Resultado da Soma (A + B) ---");
          console.log(formatarMatriz(resultado, "A + B"));
        } catch (err) {
          console.log(`\n>> Erro de Operação: ${err.message}`);
        }
      }

    } else if (opcao === "6") {
      if (!matrizA || !matrizB) {
        console.log(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a subtração.");
      } else {
        try {
          const resultado = subtrairMatrizes(matrizA, matrizB);
          console.log("\n--- Resultado da Subtração (A - B) ---");
          console.log(formatarMatriz(resultado, "A - B"));
        } catch (err) {
          console.log(`\n>> Erro de Operação: ${err.message}`);
        }
      }

    } else if (opcao === "7") {
      if (!matrizA || !matrizB) {
        console.log(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a multiplicação.");
      } else {
        try {
          const resultado = multiplicarMatrizes(matrizA, matrizB);
          console.log("\n--- Resultado do Produto (A x B) ---");
          console.log(formatarMatriz(resultado, "A x B"));
        } catch (err) {
          console.log(`\n>> Erro de Operação: ${err.message}`);
        }
      }

    } else if (opcao === "8") {
      const selecao = await selecionarMatriz(matrizA, matrizB);
      if (selecao) {
        const { matriz, nome } = selecao;
        const k = await lerFloat(`Informe o escalar (número real) para multiplicar a Matriz ${nome}: `);
        const resultado = multiplicarEscalar(matriz, k);
        console.log(`\n--- Resultado da Multiplicação (${k} * ${nome}) ---`);
        console.log(formatarMatriz(resultado, `${k} * ${nome}`));
      }

    } else if (opcao === "9") {
      const selecao = await selecionarMatriz(matrizA, matrizB);
      if (selecao) {
        const { matriz, nome } = selecao;
        const resultado = transporMatriz(matriz);
        console.log(`\n--- Matriz Transposta de ${nome} (${nome}^T) ---`);
        console.log(formatarMatriz(resultado, `${nome}^T`));
      }

    } else if (opcao === "10") {
      const selecao = await selecionarMatriz(matrizA, matrizB);
      if (selecao) {
        const { matriz, nome } = selecao;
        const [linhas, colunas] = obterDimensoes(matriz);
        console.log(`\n--- Análise de Propriedades da Matriz ${nome} (${linhas}x${colunas}) ---`);

        // Quadrada
        const ehQuad = verificarQuadrada(matriz);
        console.log(`1. Matriz Quadrada: ${ehQuad ? "SIM" : "NÃO"}`);

        // Simétrica
        const { ehSimetrica, mensagem: msgSim } = verificarSimetrica(matriz);
        console.log(`2. Matriz Simétrica: ${ehSimetrica ? "SIM" : "NÃO"} (${msgSim})`);

        // Identidade
        const { ehIdentidade, mensagem: msgId } = verificarIdentidade(matriz);
        console.log(`3. Matriz Identidade: ${ehIdentidade ? "SIM" : "NÃO"} (${msgId})`);
      }

    } else if (opcao === "11") {
      const selecao = await selecionarMatriz(matrizA, matrizB);
      if (selecao) {
        const { matriz, nome } = selecao;
        try {
          const traco = calcularTraco(matriz);
          console.log(`\n--- Traço da Matriz ${nome} ---`);
          console.log(`Tr(${nome}) = ${traco}`);
        } catch (err) {
          console.log(`\n>> Erro: ${err.message}`);
        }
      }

    } else if (opcao === "12") {
      const selecao = await selecionarMatriz(matrizA, matrizB);
      if (selecao) {
        const { matriz, nome } = selecao;
        try {
          const det = calcularDeterminante(matriz);
          console.log(`\n--- Determinante da Matriz ${nome} ---`);
          console.log(`det(${nome}) = ${det}`);
        } catch (err) {
          console.log(`\n>> Erro: ${err.message}`);
        }
      }

    } else if (opcao === "0") {
      console.log("\nEncerrando a Calculadora de Matrizes. Até logo!");
      rl.close();
      process.exit(0);

    } else {
      console.log(">> Opção inválida! Por favor, digite um número correspondente ao menu.");
    }
  }
}

// Inicia execução
main().catch((err) => {
  console.error("Erro inesperado:", err);
  rl.close();
  process.exit(1);
});

File: python/calculadora.py
"""
Programa: calculadora.py
Descrição: Interface de console interativa para a Calculadora de Matrizes.
Disciplina: Álgebra Linear — Fatec Osasco
Curso: Desenvolvimento de Software Multiplataforma
"""

import sys
from typing import Optional
from core_matrizes import (
    Matriz,
    obter_dimensoes,
    criar_matriz_vazia,
    gerar_matriz_aleatoria,
    formatar_matriz,
    somar_matrizes,
    subtrair_matrizes,
    multiplicar_escalar,
    transpor_matriz,
    multiplicar_matrizes,
    verificar_quadrada,
    verificar_simetrica,
    verificar_identidade,
    calcular_traco,
    calcular_determinante
)


def ler_inteiro_positivo(mensagem: str) -> int:
    """Lê um número inteiro estritamente positivo (> 0) com validação de entrada."""
    while True:
        try:
            entrada = input(mensagem).strip()
            valor = int(entrada)
            if valor > 0:
                return valor
            print(">> Erro: O valor deve ser um número inteiro maior que zero.")
        except ValueError:
            print(">> Erro: Entrada inválida. Por favor, digite um número inteiro válido.")


def ler_float(mensagem: str) -> float:
    """Lê um número decimal (float) com validação de entrada, aceitando ponto ou vírgula."""
    while True:
        try:
            entrada = input(mensagem).strip().replace(',', '.')
            return float(entrada)
        except ValueError:
            print(">> Erro: Entrada inválida. Por favor, digite um número real válido.")


def cadastrar_matriz_fluxo(nome: str) -> Matriz:
    """
    Fluxo guiado para cadastro de uma matriz (A ou B).
    Permite escolha entre entrada manual elemento a elemento ou geração aleatória.
    """
    print(f"\n--- Cadastro da Matriz {nome} ---")
    linhas = ler_inteiro_positivo(f"Informe o número de linhas de {nome}: ")
    colunas = ler_inteiro_positivo(f"Informe o número de colunas de {nome}: ")

    print("\nEscolha o método de preenchimento:")
    print("  1 - Manual (digitar elemento por elemento)")
    print("  2 - Automático (valores inteiros aleatórios)")

    while True:
        opcao = input("Opção desejada [1 ou 2]: ").strip()
        if opcao == "1":
            matriz = criar_matriz_vazia(linhas, colunas)
            print(f"\nDigite os elementos da matriz {nome} ({linhas}x{colunas}):")
            for i in range(linhas):
                for j in range(colunas):
                    matriz[i][j] = ler_float(f"  Elemento [{i + 1}, {j + 1}]: ")
            print(f"\nMatriz {nome} cadastrada com sucesso!")
            print(formatar_matriz(matriz, nome))
            return matriz
        elif opcao == "2":
            min_val = int(ler_float("Valor mínimo para os números aleatórios [padrão -10]: ") or -10)
            max_val = int(ler_float("Valor máximo para os números aleatórios [padrão 10]: ") or 10)
            matriz = gerar_matriz_aleatoria(linhas, colunas, min_val, max_val)
            print(f"\nMatriz {nome} gerada aleatoriamente com sucesso!")
            print(formatar_matriz(matriz, nome))
            return matriz
        else:
            print(">> Opção inválida! Escolha 1 para Manual ou 2 para Automático.")


def selecionar_matriz(matriz_a: Optional[Matriz], matriz_b: Optional[Matriz]) -> Optional[tuple[Matriz, str]]:
    """Permite ao usuário escolher entre Matriz A ou Matriz B para operações unárias."""
    print("\nQual matriz deseja utilizar?")
    print("  A - Matriz A")
    print("  B - Matriz B")
    escolha = input("Escolha [A/B]: ").strip().upper()
    if escolha == "A":
        if matriz_a is None:
            print(">> Atenção: A Matriz A ainda não foi cadastrada.")
            return None
        return matriz_a, "A"
    elif escolha == "B":
        if matriz_b is None:
            print(">> Atenção: A Matriz B ainda não foi cadastrada.")
            return None
        return matriz_b, "B"
    else:
        print(">> Opção inválida! Escolha 'A' ou 'B'.")
        return None


def menu_principal():
    """Loop principal de execução da calculadora de matrizes."""
    matriz_a: Optional[Matriz] = None
    matriz_b: Optional[Matriz] = None

    while True:
        print("\n" + "=" * 45)
        print("         CALCULADORA DE MATRIZES")
        print("=" * 45)
        print(" 1  - Cadastrar Matriz A")
        print(" 2  - Cadastrar Matriz B")
        print(" 3  - Exibir Matriz A")
        print(" 4  - Exibir Matriz B")
        print(" 5  - Somar (A + B)")
        print(" 6  - Subtrair (A - B)")
        print(" 7  - Multiplicar (A x B)")
        print(" 8  - Multiplicar Matriz por Escalar")
        print(" 9  - Transpor Matriz")
        print("--- [Opcionais / Bônus] ---")
        print(" 10 - Verificar Propriedades (Quadrada, Simétrica, Identidade)")
        print(" 11 - Calcular Traço da Matriz")
        print(" 12 - Calcular Determinante")
        print(" 0  - Sair")
        print("=" * 45)

        opcao = input("Digite a opção desejada: ").strip()

        if opcao == "1":
            matriz_a = cadastrar_matriz_fluxo("A")

        elif opcao == "2":
            matriz_b = cadastrar_matriz_fluxo("B")

        elif opcao == "3":
            if matriz_a is None:
                print(">> Matriz A não cadastrada. Utilize a opção 1 primeiro.")
            else:
                print("\n" + formatar_matriz(matriz_a, "A"))

        elif opcao == "4":
            if matriz_b is None:
                print(">> Matriz B não cadastrada. Utilize a opção 2 primeiro.")
            else:
                print("\n" + formatar_matriz(matriz_b, "B"))

        elif opcao == "5":
            if matriz_a is None or matriz_b is None:
                print(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a soma.")
            else:
                try:
                    resultado = somar_matrizes(matriz_a, matriz_b)
                    print("\n--- Resultado da Soma (A + B) ---")
                    print(formatar_matriz(resultado, "A + B"))
                except ValueError as err:
                    print(f"\n>> Erro de Operação: {err}")

        elif opcao == "6":
            if matriz_a is None or matriz_b is None:
                print(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a subtração.")
            else:
                try:
                    resultado = subtrair_matrizes(matriz_a, matriz_b)
                    print("\n--- Resultado da Subtração (A - B) ---")
                    print(formatar_matriz(resultado, "A - B"))
                except ValueError as err:
                    print(f"\n>> Erro de Operação: {err}")

        elif opcao == "7":
            if matriz_a is None or matriz_b is None:
                print(">> Erro: Ambas as matrizes (A e B) devem estar cadastradas para realizar a multiplicação.")
            else:
                try:
                    resultado = multiplicar_matrizes(matriz_a, matriz_b)
                    print("\n--- Resultado do Produto (A x B) ---")
                    print(formatar_matriz(resultado, "A x B"))
                except ValueError as err:
                    print(f"\n>> Erro de Operação: {err}")

        elif opcao == "8":
            selecao = selecionar_matriz(matriz_a, matriz_b)
            if selecao:
                matriz, nome = selecao
                k = ler_float(f"Informe o escalar (número real) para multiplicar a Matriz {nome}: ")
                resultado = multiplicar_escalar(matriz, k)
                print(f"\n--- Resultado da Multiplicação ({k} * {nome}) ---")
                print(formatar_matriz(resultado, f"{k} * {nome}"))

        elif opcao == "9":
            selecao = selecionar_matriz(matriz_a, matriz_b)
            if selecao:
                matriz, nome = selecao
                resultado = transpor_matriz(matriz)
                print(f"\n--- Matriz Transposta de {nome} ({nome}^T) ---")
                print(formatar_matriz(resultado, f"{nome}^T"))

        elif opcao == "10":
            selecao = selecionar_matriz(matriz_a, matriz_b)
            if selecao:
                matriz, nome = selecao
                linhas, colunas = obter_dimensoes(matriz)
                print(f"\n--- Análise de Propriedades da Matriz {nome} ({linhas}x{colunas}) ---")
                
                # Quadrada
                eh_quadrada = verificar_quadrada(matriz)
                print(f"1. Matriz Quadrada: {'SIM' if eh_quadrada else 'NÃO'}")
                
                # Simétrica
                eh_simetrica, msg_sim = verificar_simetrica(matriz)
                print(f"2. Matriz Simétrica: {'SIM' if eh_simetrica else 'NÃO'} ({msg_sim})")
                
                # Identidade
                eh_identidade, msg_id = verificar_identidade(matriz)
                print(f"3. Matriz Identidade: {'SIM' if eh_identidade else 'NÃO'} ({msg_id})")

        elif opcao == "11":
            selecao = selecionar_matriz(matriz_a, matriz_b)
            if selecao:
                matriz, nome = selecao
                try:
                    traco = calcular_traco(matriz)
                    print(f"\n--- Traço da Matriz {nome} ---")
                    print(f"Tr({nome}) = {traco if traco != int(traco) else int(traco)}")
                except ValueError as err:
                    print(f"\n>> Erro: {err}")

        elif opcao == "12":
            selecao = selecionar_matriz(matriz_a, matriz_b)
            if selecao:
                matriz, nome = selecao
                try:
                    det = calcular_determinante(matriz)
                    print(f"\n--- Determinante da Matriz {nome} ---")
                    print(f"det({nome}) = {det if det != int(det) else int(det)}")
                except ValueError as err:
                    print(f"\n>> Erro: {err}")

        elif opcao == "0":
            print("\nEncerrando a Calculadora de Matrizes. Até logo!")
            sys.exit(0)

        else:
            print(">> Opção inválida! Por favor, digite um número correspondente ao menu.")


if __name__ == "__main__":
    try:
        menu_principal()
    except KeyboardInterrupt:
        print("\n\nExecução interrompida pelo usuário. Saindo...")
        sys.exit(0)
